src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const stateCopy = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const removeKey of action.keysToRemove) {
+          delete stateCopy[removeKey];
+        }
+        break;
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
+        break;
+    }
+
+    stateHistory.push({ ...stateCopy });
```

Comment:
Pushing a copy of the state object to 'stateHistory' is correctly done outside of the switch block, which is good practice as it avoids repetition.

Reply:
no comments...